src/libpmemfile-posix/write.c

```diff
@@ -164,55 +191,101 @@ pmemfile_pwritev_internal(PMEMfilepool *pfp,
+
+		if (error)
+			goto end;
+	}
+
+	struct pmemfile_time tm;
+	if (get_current_time(&tm)) {
+		error = errno;
+		goto end;
+	}
 
-		for (int i = 0; i < iovcnt; ++i) {
-			size_t len = iov[i].iov_len;
+	/*
+	 * We have to update mtime before actually modifying file contents,
+	 * just in case of crash/power failure.
+	 */
+	bool mtime_slot = inode_next_slot(inode, mtime);
+	inode->mtime[mtime_slot] = tm;
+	pmemfile_persist(pfp, &inode->mtime[mtime_slot]);
 
-			if ((pmemfile_ssize_t)len < 0)
-				len = SSIZE_MAX;
+	inode->slots.mtime = mtime_slot;
+	pmemfile_persist(pfp, &inode->slots);
 
-			if ((pmemfile_ssize_t)(ret + len) < 0)
-				len = SSIZE_MAX - ret;
+	for (int i = 0; i < iovcnt; ++i) {
+		size_t len = iov[i].iov_len;
 
-			if (offset + len < offset) /* overflow check */
-				len = SIZE_MAX - offset;
+		if ((pmemfile_ssize_t)len < 0)
+			len = SSIZE_MAX;
 
-			if (len > 0)
-				vinode_write(pfp, vinode, offset, last_block,
-						iov[i].iov_base, len);
+		if ((pmemfile_ssize_t)(ret + len) < 0)
+			len = SSIZE_MAX - ret;
 
-			ret += len;
-			offset += len;
+		if (offset + len < offset) /* overflow check */
+			len = SIZE_MAX - offset;
 
-			if (len != iov[i].iov_len)
-				break;
-		}
+		if (len > 0)
+			vinode_write(pfp, vinode, offset, last_block,
+					iov[i].iov_base, len);
 
-		/*
-		 * Update metadata only when any of the buffer lengths
-		 * was != 0.
-		 */
-		if (ret > 0) {
-			struct pmemfile_time tm;
-			tx_get_current_time(&tm);
-
-			if (offset > inode_get_size(inode)) {
-				inode_tx_set_size(inode, offset);
-				inode_tx_set_ctime(inode, tm);
-			}
-
-			inode_tx_set_allocated_space(inode, allocated_space);
-			inode_tx_set_mtime(inode, tm);
-		}
-	} TX_ONABORT {
-		if (errno == ENOMEM)
-			errno = ENOSPC;
+		ret += len;
+		offset += len;
+
+		if (len != iov[i].iov_len)
+			break;
+	}
+	ASSERT(ret > 0);
+
+	struct pmemfile_time starttm = tm;
+	if (get_current_time(&tm)) {
 		error = errno;
-		vinode_restore_on_abort(vinode);
-	} TX_END
+		goto end;
+	}
+
+	int64_t tm_diff = (tm.sec - starttm.sec) * 1000000000 +
+			tm.nsec - starttm.nsec;
+	bool slots_changed = false;
+
+	/*
+	 * In theory we don't have to do it at all. We have a write lock on
+	 * vinode, so file changes can't be observed by another thread.
+	 * However if we performed long write which took more than, let's say,
+	 * 1ms we can update mtime. Time taken to do this will be negligible
+	 * comparing to time spent on copying data.
+	 */
+	if (tm_diff >= 1000000) {
+		mtime_slot = inode_next_slot(inode, mtime);
+		inode->mtime[mtime_slot] = tm;
+		pmemfile_persist(pfp, &inode->mtime[mtime_slot]);
+
+		inode->slots.mtime = mtime_slot;
+		slots_changed = true;
+	}
+
+	if (offset > inode_get_size(inode)) {
+		bool size_slot = inode_next_slot(inode, size);
+		inode->size[size_slot] = offset;
+		pmemfile_flush(pfp, &inode->size[size_slot]);
+
+		bool ctime_slot = inode_next_slot(inode, ctime);
+		inode->ctime[ctime_slot] = tm;
```

Comment:
Does this mean, if `tm_diff < 1000000` after appending a file, we update ctime, but not update mtime?
Specifically, can't this misleading some user into thinking some file permissions have been modified, since `ctime > mtime` ?

Reply:
Perhaps this line could rather be:
`inode->ctime[ctime_slot] = inode->mtime[ctime_slot]`

Reply:
I added mtime update in tm_diff<1000000 case when ctime is modified.

WRT the 2nd comment:
You can't use ctime_slot for reading mtime, but assuming it's a typo and you meant mtime_slot - it's not that great idea because we would read recently flushed data, which means cpu stall...

Reply:
BTW, I simplified this code in the latest version.